File: companies/Wayfair/253.meeting-rooms-ii.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        /*
        Approach: sort meetings by start time. A min-heap stores end times of
        currently occupied rooms. Before placing a meeting, free every room whose
        end time is <= the meeting start. The maximum heap size is the answer.

        C++ notes: priority_queue<int, vector<int>, greater<int>> is a min-heap.
        Complexity: O(n log n) time, O(n) space.
        */
        if (intervals.empty()) return 0;
        sort(intervals.begin(), intervals.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[0] < b[0];
        });
        priority_queue<int, vector<int>, greater<int>> heap;
        int best = 0;
        for (auto& interval : intervals) {
            int start = interval[0], end = interval[1];
            while (!heap.empty() && heap.top() <= start) heap.pop();
            heap.push(end);
            best = max(best, (int)heap.size());
        }
        return best;
    }
};

File: companies/Compass/253.meeting-rooms-ii.go
package leetcode

import (
	"container/heap"
	"sort"
)

type endHeap253 []int

func (h endHeap253) Len() int           { return len(h) }
func (h endHeap253) Less(i, j int) bool { return h[i] < h[j] }
func (h endHeap253) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *endHeap253) Push(x any)        { *h = append(*h, x.(int)) }
func (h *endHeap253) Pop() any          { old := *h; x := old[len(old)-1]; *h = old[:len(old)-1]; return x }

// MinMeetingRooms253 sorts meetings by start and keeps a min-heap of active end
// times. Meetings whose end <= current start free a room; heap size is rooms in
// use, and the maximum size is the answer.
//
// Time: O(n log n). Space: O(n).
func MinMeetingRooms253(intervals [][]int) int {
	if len(intervals) == 0 {
		return 0
	}
	sort.Slice(intervals, func(i, j int) bool { return intervals[i][0] < intervals[j][0] })
	h := &endHeap253{}
	heap.Init(h)
	best := 0
	for _, in := range intervals {
		start, end := in[0], in[1]
		for h.Len() > 0 && (*h)[0] <= start {
			heap.Pop(h)
		}
		heap.Push(h, end)
		best = maxInt(best, h.Len())
	}
	return best
}
